### ISSUE TITLE
Support loading TSM files via `influxdb_iox write`

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#8037]
__Use case:__
We would like to help customers begin the migration process from TSM to IOx. To do so they need a way to export the data that is currently stored in TSM (both the OSS version as well as the one used in Cloud 2) into IOx. 

The larger / longer term plan and design is described in the document [TSM => IOx Bulk Data Export in Cloud 2 MT](https://docs.google.com/document/d/1APUsypAGAzZCjdxOKB6aolNohbzVu4v4uVjHl56WMAw/edit) which is efficient and scalable and will not involve a conversion to line protocol. 

However, before that system is complete, it would be valuable to offer a less efficient way to load data stored in TSM files into IOx. 

Currently IOx supports loading data via the `influxdb_iox` CLI by taking line protocol, chunking it, and then uploading it in parallel to the `/api/v2/write` endpoint.

Also, IOx has a TSM reader written by @e-dard  in https://github.com/influxdata/influxdb_iox/tree/main/influxdb_tsm

__Proposal:__
 I propose extending the existing `influxdb_iox write` functionality to directly load TSM files into IOx (via the `/api/v2/write` API).

```shell
# (works today) Write data in gzipped line protocol file
influxdb_iox write my_namespace foo.lp.gz

# (works today) write data from a (IOx) created parquet file
influxdb_iox write my_namespace foo.parquet

# (PROPOSAL) write data from a TSM file
influxdb_iox write my_namespace foo.tsm
```

`influxdb_iox write my_namespace foo.tsm`, would:
1. Open the `foo.tsm` file using [influxdb_tsm](https://github.com/influxdata/influxdb_iox/tree/main/influxdb_tsm)
2. Convert the relevant series into line protocol (similarly to how [parquet_to_line_protocol](https://github.com/influxdata/influxdb_iox/tree/main/parquet_to_line_protocol) converts data in parquet back to line protocol using the [`LineProtocolBuilder`](https://github.com/influxdata/influxdb_iox/blob/main/influxdb_line_protocol/src/builder.rs#L105)
3. Send the line protocol to the /api/v2/write endpoint

I made a basic sketch of how this would work  in https://github.com/influxdata/influxdb_iox/pull/7418

__Context__
Describe other solutions or features you considered.

**Comment by btasker** [comment]
> The biggest pain point is identifying and handling inter-shard type mismatches and any schema changes required for migrations. 

There is a second pain point, which this proposal might well help with.

TSM -> LP tends to be performed using `influx_inspect export`. It does its job very well, but *it's single threaded*, so it can be very low to dump large datasets.

We do have ways of mitigating that (for example, by splitting files out and running multiple `influx_inspect`'s), but it would probably make things **much** more time efficient if `influxdb_iox write` were able to accept a 1.x `data` dir and import TSMs in parallel.

Where a migration process involves dump to LP (import into C2 TSM takes a different route) it looks a little like this

- use `influx_inspect` to dump to LP (can be parallelized, but still slow)
- split files into many files (is as fast as your box can handle)
- Parallelize files with `find ./ -name '*.tsm' | xargs .... etc` (can be **very** fast)

I'd imagine it'd fall outside the scope of this issue, but a related "nice-to-have" would be for `influxdb_iox` to accept an (optional) file mapping fields to type, so that type conflicts between shards can be resolved automatically by coercing (or attempting to) to the specified type.

For Schema changes, my 0.02 is that I think it'd probably be reasonable to say that they need to be made beforehand (perhaps by running Flux to transform into a new DB and then providing the TSMs for that).

**Comment by alamb** [comment]
> We do have ways of mitigating that (for example, by splitting files out and running multiple influx_inspect's), but it would probably make things much more time efficient if influxdb_iox write were able to accept a 1.x data dir and import TSMs in parallel.

`influxdb_iox`  supports loading multiple files passed on the command line -- I think it sends up to 10 requests in parallel. It is quite good at keeping the CPU busy

So in my idealized world, what would happen is someone would rin

```
influxdb_iox write namespace_name /path/to/tsm/*.tsm
```

Which would then convert to line protocol (incrementally, in memory) and send it to IOx. 

I did not appreciate the "recover mid-export" as well as type conflicts between shards -- 

> For Schema changes, my 0.02 is that I think it'd probably be reasonable to say that they need to be made beforehand (perhaps by running Flux to transform into a new DB and then providing the TSMs for that).

**Comment by 8none1** [comment]
I have a slightly different opinion...

I think `influxdb_iox write` should stick to writing LP files and `influx_inspect` should gain the ability to export to LP which is appropriately shaped for ingestion in to IOx (and the tool could become multi threaded).  

The data in TSM (be it OSS or 1.X Enterprise) will not be appropriate for ingestion in to IOx without sanity checking, type normalisation and a review of fields and tags.

I would prefer to put this functionality in to `influx_inspect` which would then be able to produce LP which is ready to be passed in to `influxdb_iox write`.  

The downside of this approach is the LP files will be massive.  Simply gzipping them on the way out might be enough.

I think that @pauldix has some views on this too, so cc'ing for visibility.

Regarding the priority of this in the overall TSM -> IOx data migration plan (linked to by Andrew above); my expectation is (and this has not been sufficiently analysed yet to be considered correct) that the backend Parquet importer/adder to the catalogue could be a large chunk of work.
We also have to write the front end which will ingest TSM files and output "checked" Parquet ready for upload - this is also a decent chunk of work. 
So.. if extending existing tooling can be done in a number of weeks, then I think we can use this as a stop-gap to enable early migrations (probably hand-held by the support team) for POCs etc while the main system is designed and built.
But I would want to discuss this with the write team first.

Adding @jacobmarble and @stuartcarnie for visibility.
